### PR TITLE
feat: add support for the TCP_USER_TIMEOUT option

### PR DESCRIPTION
**Pull request by mildsunrise**
keep-alive is usually wanted for dead peer detection. But keep-alive is only one half of that; if the peer dies before acknowledging data, the `TCP_USER_TIMEOUT` option is what decides when the connection gets closed, and that's usually set to 10 minutes. (correct me if wrong)

So I suggest we add support for setting the TCP User Timeout on a socket, too. Currently only implemented for Linux, I haven't had time to look this on other platforms yet.

### CODE REVIEW

**Review by hertzg — COMMENTED** [review]
Amazing work @mildsunrise with preparing PR 👏 . 

> So I suggest we add support for setting the TCP User Timeout on a socket, too.

I have not used `TCP_USER_TIMEOUT` before and after reading a bit about it seems like it would complement the rest of the module functinality very well 👍. 

> Currently only implemented for Linux, I haven't had time to look this on other platforms yet.

So far what I found is that the [rfc5482](https://tools.ietf.org/html/rfc5482) might be considered "young(ish)" and seems to be either linux specific or just not yet implemented in other kern. 
It's unsupported on [`freebsd`](https://wiki.freebsd.org/TransportProtocols/tcp_rfc_compliance) and `darwin` (don't have any strong facts for macos, only few words here and there) but apparently supported on [`win32`](https://stackoverflow.com/a/12948084). 

- [ ] Therefore I suggested throwing in `set` and `get` methods is they are used on those platforms. I'm open to suggestions here.

screenshot of freebsd tcp rfc compliance wiki page listing RFC 5482 as unspported

It think it would be great to have an end to end test confirming that it does what it's supposed to do. A test similar to [`test-tcpdump.js `](https://github.com/hertzg/node-net-keepalive/blob/master/test/integration/test-tcpdump.js) but only checking that `TCP_USER_TIMEOUT` does what it's expected to do on `linux.

A [great drill-down by Couldflare](https://blog.cloudflare.com/when-tcp-sockets-refuse-to-die/) regarding this (and many other cases) demonstrates that it could be visualised using `tcpdump`.

- [ ] Can we create an e2e test using tcpdump to test that it was actually set and at least picked up by kernel?

**Comment by hertzg** [comment]
For now you can ignore the failing check for node `v12`, It's tracked in #93 and if other versions are passing we can ignore it.

File: lib/index.js [hunk 1]
```diff
@@ -213,3 +213,96 @@ module.exports.getKeepAliveProbes = function getKeepAliveProbes(socket) {
...
+ *
+ * Note: The msec will be rounded towards the closest integer
+ *
+ * @since v1.4.0
+ * @param {Net.Socket} socket to set the value for
+ * @param {number} msecs to wait for unacknowledged data before closing the connection
+ *
+ * @returns {boolean} <code>true</code> on success
+ *
+ * @example <caption>Set user timeout to 30 seconds (<code>1000</code> milliseconds) for socket <code>s</code></caption>
```
[the remark this answers is not in the document]
Comment (mildsunrise) [reply lib/index.js:228]: it was accidental, thanks!

[the remark this answers is not in the document]
Comment (mildsunrise) [reply lib/index.js:228]: it was accidental, thanks!

**Comment by mildsunrise** [comment]
Thanks for reviewing :)

>  Can we create an e2e test using tcpdump to test that it was actually set and at least picked up by kernel?

Okay, I'll try to implement an e2e test using tcpdump, I think it should be doable.

> So far what I found is that the rfc5482 might be considered "young(ish)" and seems to be either linux specific or just not yet implemented in other kernels.
> It's unsupported on freebsd and darwin (don't have any strong facts for macos, only few words here and there) but apparently supported on win32.

Not exactly... TCP itself defines the user timeout. What RFC5482 does is add an option to the protocol so that you can _advertise_ the timeout to the other peer, rather than just setting it locally.

It's not clear how (or if) the `TCP_USER_TIMEOUT` option can be used with other kernels... FreeBSD seems to point to `SO_SNDTIMEO`, which is _similar_ but I'm not sure it does the same since it apparently only affects blocking syscalls? So I'm not sure at this point.

>  Therefore I suggested throwing in set and get methods if they are called on unsupported platforms. I'm open to suggestions here.

Absolutely!

File: lib/index.js [hunk 2]
```diff
@@ -213,3 +213,96 @@ module.exports.getKeepAliveProbes = function getKeepAliveProbes(socket) {
...
+    'setUserTimeout expects an instance of socket as its first argument'
+  )
+  Assert.strictEqual(
+    msecs != null ? msecs.constructor : void 0,
+    Number,
+    'setUserTimeout requires msec to be a Number'
+  )
+
+  const fd = _getSocketFD(socket),
+    seconds = ~~msecs,
```
[the remark this answers is not in the document]
Comment (mildsunrise) [reply lib/index.js]: Applied, thanks!

**Comment by hertzg** [comment]
> Okay, I'll try to implement an e2e test using tcpdump, I think it should be doable.

That is amazing! I believe we could use some pieces from [repo for cloud-flare blog post](https://github.com/cloudflare/cloudflare-blog/tree/master/2019-09-tcp-keepalives):

* [tcpdump flags and filters to produce the same output as in the blog post](https://github.com/cloudflare/cloudflare-blog/blob/master/2019-09-tcp-keepalives/utils.py#L55)
* [snippet of iptables flags to drop packets](https://github.com/cloudflare/cloudflare-blog/blob/master/2019-09-tcp-keepalives/utils.py#L43) 
* [user_timeout full example](https://github.com/cloudflare/cloudflare-blog/blob/master/2019-09-tcp-keepalives/test-idle.py)

> Not exactly... TCP itself defines the user timeout. What RFC5482 does is add an option to the protocol so that you can advertise the timeout to the other peer, rather than just setting it locally.

Ah! I see... Thanks for the explanation. 🙏

> It's not clear how (or if) the TCP_USER_TIMEOUT option can be used with other kernels... FreeBSD seems to point to SO_SNDTIMEO, which is similar but I'm not sure it does the same since it apparently only affects blocking syscalls? So I'm not sure at this point.

I'm OK starting with `linux` only support, given that we document that fact as well as throw in other platforms so that users would know it.

Later we can see if there is a demand for it on other platforms.

**Comment by hertzg** [comment]
@mildsunrise Gentle ping, should I take over or do you have some progress?

**Comment by mildsunrise** [comment]
you can take over :)

**Comment by hertzg** [comment]
@all-contributors please add @mildsunrise for code, docs, tests

**Pull request merged**